aws/resource_aws_route.go

```diff
@@ -382,186 +277,234 @@ func resourceAwsRouteRead(d *schema.ResourceData, meta interface{}) error {
+func routeDestinationAndTargetAttributes(d *schema.ResourceData) (string, string, error) {
+	destinationAttrs := map[string]struct {
+		ipv4Destination bool
+		ipv6Destination bool
+	}{
+		"destination_cidr_block":      {true, false},
+		"destination_ipv6_cidr_block": {false, true},
+	}
 
-	findOpts := &ec2.DescribeRouteTablesInput{
-		RouteTableIds: []*string{&routeTableID},
+	destinationAttr := ""
+	ipv4Destination := false
+	ipv6Destination := false
+	for attr, kind := range destinationAttrs {
+		if v := d.Get(attr).(string); v != "" {
+			if destinationAttr != "" {
+				return "", "", fmt.Errorf("`%s` conflicts with `%s`", attr, destinationAttr)
+			}
+
+			destinationAttr = attr
+			ipv4Destination = kind.ipv4Destination
+			ipv6Destination = kind.ipv6Destination
+		}
 	}
 
-	resp, err := conn.DescribeRouteTables(findOpts)
-	if err != nil {
-		return nil, err
+	if destinationAttr == "" {
+		keys := []string{}
+		for k := range destinationAttrs {
+			keys = append(keys, k)
+		}
+
+		return "", "", fmt.Errorf("one of `%s` must be specified", strings.Join(keys, "`, `"))
 	}
 
-	if len(resp.RouteTables) < 1 || resp.RouteTables[0] == nil {
-		return nil, nil
+	targetAttrs := map[string]struct {
+		ipv4Destination bool
+		ipv6Destination bool
+	}{
+		"egress_only_gateway_id":    {false, true},
+		"gateway_id":                {true, true},
+		"instance_id":               {true, true},
+		"nat_gateway_id":            {true, false},
+		"network_interface_id":      {true, true},
+		"transit_gateway_id":        {true, false},
+		"vpc_peering_connection_id": {true, true},
 	}
 
-	if cidr != "" {
-		for _, route := range (*resp.RouteTables[0]).Routes {
-			if route.DestinationCidrBlock != nil && *route.DestinationCidrBlock == cidr {
-				return route, nil
+	targetAttr := ""
+	for attr, allowed := range targetAttrs {
+		if v := d.Get(attr).(string); v != "" && d.HasChange(attr) {
+			if targetAttr != "" {
+				return "", "", fmt.Errorf("`%s` conflicts with `%s`", attr, targetAttr)
+			}
+
+			if (ipv4Destination && !allowed.ipv4Destination) || (ipv6Destination && !allowed.ipv6Destination) {
+				return "", "", fmt.Errorf("`%s` not allowed for `%s` target", destinationAttr, attr)
 			}
+
+			targetAttr = attr
 		}
+	}
 
-		return nil, nil
+	if targetAttr == "" {
+		keys := []string{}
+		for k := range targetAttrs {
+			keys = append(keys, k)
+		}
+
+		return "", "", fmt.Errorf("one of `%s` must be specified", strings.Join(keys, "`, `"))
 	}
 
-	if ipv6cidr != "" {
-		for _, route := range (*resp.RouteTables[0]).Routes {
-			if cidrBlocksEqual(aws.StringValue(route.DestinationIpv6CidrBlock), ipv6cidr) {
-				return route, nil
-			}
+	return destinationAttr, targetAttr, nil
+}
+
+// createRoute attempts to create a route.
+// The specified eventual consistency timeout is respected.
+// Any error is returned.
+func createRoute(conn *ec2.EC2, input *ec2.CreateRouteInput, timeout time.Duration) error {
```

Comment:
Can be moved to internal `waiter` package once https://github.com/terraform-providers/terraform-provider-aws/pull/13036 is merged.